### PR TITLE
Manually sync reviews

### PR DESCRIPTION
**Pull request by TsankoBg**
https://app.clubhouse.io/webwinkelkeur/story/5556/add-button-for-manually-syncing-product-reviews-in-woocommerce

Changes made: 
 - add a button that makes an ajax call to trigger review sync
 - indicate if manual sync was successful 
 - add last sync and next sync date 
 - minor improvements to front-end

[ch5556]

### CODE REVIEW

File: common/src/BasePlugin.php [hunk 1]
```diff
@@ -218,4 +218,29 @@ private function isUniqueAttribute(array $array, string $value): bool {
...
+        return $this->getReviewSyncDate(strtotime(
+            get_option($this->getOptionName('last_synced'))
+        ));
+    }
+
+    private function getReviewSyncDate($date): string {
+        if ($date) {
+            return htmlentities(date("Y-m-d H:i:s", $date));
+        }
+        return 'Not registered.';
```
Comment (apeschar) [inline common/src/BasePlugin.php]: This should be translated.

File: common/src/BasePlugin.php [hunk 2]
```diff
@@ -218,4 +218,29 @@ private function isUniqueAttribute(array $array, string $value): bool {
     private function isValidGtin(string $value): bool {
         return preg_match('/^\d{8}(?:\d{4,6})?$/', $value);
     }
+
+    public function getManualSyncAction(): string {
```
Comment (apeschar) [inline common/src/BasePlugin.php]: Since all of these functions are specific to the WooCommerce integration, don't they belong in that class?

File: common/src/WooCommerce.php [hunk 3]
```diff
@@ -282,6 +283,13 @@ private function get_product_data(array $order_arr) {
         return $products;
     }
 
+    public function manualReviewSync() {
+        check_ajax_referer($this->plugin->getManualSyncNonce());
+        $this->syncReviews();
+        echo json_encode(['status' => true]);
```
Comment (apeschar) [inline common/src/WooCommerce.php]: You should send a header `Content-Type: application/json`.

File: common/src/WooCommerce.php [hunk 4]
```diff
@@ -301,9 +309,7 @@ public function syncReviews() {
             return;
         }
         $this->processReviews($reviews);
-        if ($last_modified = (string) ($reviews[0]->modified ?? null)) {
-            update_option($this->plugin->getOptionName('last_synced'), $last_modified);
-        }
+        update_option($this->plugin->getOptionName('last_synced'), date(\DateTime::RFC3339));
```
Comment (apeschar) [inline common/src/WooCommerce.php]: No. This will miss reviews that are created between the start and end of the sync.

File: common/templates/options.php [hunk 5]
```diff
@@ -140,23 +140,35 @@
...
-                            <option value=""><?= $plugin->getActiveGtinPlugin() ? __('Automatic detection', 'webwinkelkeur') . ' (' . (explode('/', $plugin->getActiveGtinPlugin())[0] ?? '') . ')' : 'Select key'; ?></option>
-                            <optgroup label="<?= _e('Suggested keys', 'webwinkelkeur'); ?>">
-                                <?= $plugin->getSelectOptions($config['custom_gtin'], true); ?>
-                            </optgroup>
-                            <optgroup label="<?= _e('Other keys', 'webwinkelkeur'); ?>">
-                                <?= $plugin->getSelectOptions($config['custom_gtin']); ?>
-                            </optgroup>
-                        </select>
-                    </label>
+                    <label>GTIN/EAN key</label>
```
Comment (apeschar) [inline common/templates/options.php]: Why did you do this? And if you do this, this label needs a `for` attribute.

File: common/templates/options.php [hunk 6]
```diff
@@ -181,3 +193,22 @@
         <?php submit_button(); ?>
     </div>
 </form>
+<script>
+    function triggerManualSync() {
+        <?php $nonce = wp_create_nonce($plugin->getManualSyncNonce());?>
+        jQuery.ajax({
+            type: "post",
+            url: "admin-ajax.php",
+            data: {action: <?= json_encode($plugin->getManualSyncAction()); ?>, _ajax_nonce: '<?php echo $nonce; ?>'},
```
Comment (apeschar) [inline common/templates/options.php]: Use json_encode instead of echoing directly into a string.  You can also JSON encode the whole data object.

~~~
data: <?= json_encode([
  'action' => $plugin->getManualSyncAction(),
  '_ajax_nonce' => $nonce,
]); ?>

**Review by apeschar — CHANGES_REQUESTED** [review]

File: common/templates/options.php [hunk 7]
```diff
@@ -140,6 +140,20 @@
...
                     </fieldset>
+                    <fieldset>
+                        <label>
+                            <button class="button" <?= !$config['product_reviews'] ? 'disabled' : ''; ?> type="button" id="<?= $plugin->getOptionName('manual_sync_btn'); ?>" onClick="triggerManualSync()">
+                                <?= __('Sync manually', 'webwinkelkeur'); ?>
+                            </button>
+                            <span id='successful-sync' hidden style="color:#0ED826">&#10003;
+                                <?= __('Synced successfully', 'webwinkelkeur'); ?>
+                            </span>
+                        </label> </br>
```
Comment (apeschar) [inline common/templates/options.php]: What is `</br>`?  Did you mean `<br>` ?

**Review by apeschar — CHANGES_REQUESTED** [review]

File: common/templates/options.php (hunk 7 above)
Comment (TsankoBg) [reply common/templates/options.php]: yes

**Review by apeschar — APPROVED** [review]
